v2/tracing.go

```diff
@@ -5,13 +5,19 @@ import (
 
 	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
 	"github.com/Azure/go-shuttle/v2/otel"
+	"go.opentelemetry.io/otel/trace"
 )
 
 // NewTracingHandler is a shuttle middleware that extracts the context from the message Application property if available,
 // or from the existing context if not, and starts a span.
-func NewTracingHandler(next Handler) HandlerFunc {
+func NewTracingHandler(next Handler,
+	extractFn func(ctx context.Context, message *azservicebus.ReceivedMessage) (context.Context, trace.Span),
```

Comment:
this makes it a breaking change in the interface.

let's use a variadic option parameter instead, so that it does not break current consumers.

Reply:
that will require to create a TracingHandler type (and not just a func) so that the option can be applied onto something.

```go
type TracingHandler struct {
    EnrichSpan func(context.Context, trace.Span)
}
```

```go
func NewTracingHandler(next Handler, options ...func(t *TracingHandler)) HandlerFunc {
   t := &TracingHandler{}
   for _, opt in range := options {
       opt(t)
   }
   return t.Handle
}
```

then you can provide this helper:

```go
func WithSpanEnrichement(enrich func(context.Context, trace.Span) func(t *TracingHandler)) {
    return func(t *TracingHandler) {
       t.EnrichSpan = enrich
    }
}
```

Reply:
updated

Reply:
notice how the signature of the `enrich` func argument takes in a trace.Span?
this is because we let goshuttle control how the span is extracted, and only allow the user to enrich it with whatever attributes they need.

Reply:
updated in the suggested direction.